### PR TITLE
Animations: CSS var/calc polyfill calculations

### PR DESCRIPTION
**Pull request by dvoytenko**
Partial for #9129.

### CODE REVIEW

File: extensions/amp-animation/0.1/css-expr-ast.js [hunk 1]
```diff
@@ -293,4 +694,72 @@ export class CssCalcProductNode extends CssNode {
...
+          throw new Error('one of sides in multiplication must be a number');
+        }
+        multi = right.num_;
+        base = left;
+      }
+    } else {
+      if (!(right instanceof CssNumberNode)) {
+        throw new Error('denominator must be a number');
+      }
+      base = left;
```
Comment (aghassemi) [inline extensions/amp-animation/0.1/css-expr-ast.js]: also throw if right.num_ is 0? would end up being a better error message than browser provided divide by 0 error

File: extensions/amp-animation/0.1/css-expr-ast.js [hunk 2]
```diff
@@ -14,10 +14,73 @@
...
+   * @param {string} unusedVarName
+   * @return {?CssNode}
+   */
+  getVar(unusedVarName) {}
+
+  /**
+   * Returns the current font size.
+   * @return {number}
+   */
+  getCurrentFontSize() {}
```
Comment (aghassemi) [inline extensions/amp-animation/0.1/css-expr-ast.js]: I assume this would come from amp-animation's target's computedStyle later?

Comment (dvoytenko) [inline extensions/amp-animation/0.1/css-expr-ast.js]: Yes. There will be two forms: one for current target in amp-animation and one for a specifically selected target. This method address the former.

File: extensions/amp-animation/0.1/css-expr-ast.js [hunk 3]
```diff
@@ -27,6 +90,65 @@ const DATA_URL_RE = /^data\:/i;
  */
 export class CssNode {
   constructor() {}
+
+  /**
+   * A return a string CSS representation.
```
Comment (dvoytenko) [inline extensions/amp-animation/0.1/css-expr-ast.js]: Fixed

File: extensions/amp-animation/0.1/css-expr-ast.js [hunk 4]
```diff
@@ -27,6 +90,65 @@ const DATA_URL_RE = /^data\:/i;
...
+   * @return {string}
+   * @abstract
+   */
+  css() {}
+
+  /**
+   * Resolves the value of all variable components. Only performs any work if
```
Comment (dvoytenko) [inline extensions/amp-animation/0.1/css-expr-ast.js]: Thanks. Added more docs to explain. But basically, this is a public final method it provides a optimization hook to immediately resolve the node with itself if it's a constant (`isConst() == true`). And `calc()` is used to actually do calculations. Most of nodes are constants so they follow the shorter path with no additional memory allocations.

File: extensions/amp-animation/0.1/css-expr-ast.js [hunk 5]
```diff
@@ -62,37 +184,61 @@ export class CssConcatNode extends CssNode {
...
-    this.css_ = css;
+  /** @override */
+  isConst() {
+    return this.array_.reduce((acc, node) => acc && node.isConst(), true);
+  }
+
+  /** @override */
+  calc(context) {
+    let nullCount = 0;
+    const resolvedArray = this.array_.map(node => {
```
Comment (dvoytenko) [inline extensions/amp-animation/0.1/css-expr-ast.js]: Agree. Fixed.

File: extensions/amp-animation/0.1/css-expr-ast.js (hunk 1 above)
Comment (dvoytenko) [inline extensions/amp-animation/0.1/css-expr-ast.js]: I'm still investigating this part. But it appears that the right behavior is to resolve as `null`, as is done couple lines down. This, in most of cases, resolves to `none` in CSS.

File: extensions/amp-animation/0.1/css-expr-ast.js [hunk 6]
```diff
@@ -189,13 +468,50 @@ export class CssFuncNode extends CssNode {
...
+
+  /** @override */
+  css() {
+    const args = this.args_.map(node => node.css()).join(',');
+    return `${this.name_}(${args})`;
+  }
+
+  /** @override */
+  isConst() {
+    return this.args_.reduce((acc, node) => acc && node.isConst(), true);
```
Comment (jridgewell) [inline extensions/amp-animation/0.1/css-expr-ast.js]: Use `#every`.

**Pull request merged**